src/http/client.cc

```diff
@@ -19,10 +19,116 @@
  * Copyright (C) 2022 Scylladb, Ltd.
  */
 
+#include <seastar/core/loop.hh>
+#include <seastar/core/when_all.hh>
 #include <seastar/http/client.hh>
+#include <seastar/http/request.hh>
+#include <seastar/http/reply.hh>
+#include <seastar/http/response_parser.hh>
+#include <seastar/http/internal/content_source.hh>
 
 namespace seastar {
 namespace http {
 
+client::client(connected_socket&& fd)
+        : _fd(std::move(fd))
+        , _read_buf(_fd.input())
+        , _write_buf(_fd.output())
+{
+}
+
+future<> client::write_body(request& req) {
+    if (req.body_writer) {
+        return req.body_writer(internal::make_http_chunked_output_stream(_write_buf)).then([this] {
+            return _write_buf.write("0\r\n\r\n");
+        });
+    } else if (!req.content.empty()) {
+        return _write_buf.write(req.content);
+    } else {
+        return make_ready_future<>();
+    }
+}
+
+future<std::optional<reply>> client::maybe_wait_for_continue(request& req) {
+    if (req.get_header("Expect") == "") {
+        return make_ready_future<std::optional<reply>>(std::nullopt);
+    }
+
+    return _write_buf.flush().then([this] {
+        return recv_reply().then([] (reply rep) {
+            if (rep._status == reply::status_type::continue_) {
+                return make_ready_future<std::optional<reply>>(std::nullopt);
+            } else {
+                return make_ready_future<std::optional<reply>>(std::move(rep));
+            }
+        });
+    });
+}
+
+future<> client::send_request_head(request& req) {
+    if (req._version.empty()) {
+        req._version = "1.1";
+    }
+    if (req.content_length != 0) {
+        if (!req.body_writer && req.content.empty()) {
+            throw std::runtime_error("Request body writer not set and content is empty");
+        }
+        req._headers["Content-Length"] = to_sstring(req.content_length);
+    }
+
+    return _write_buf.write(req.request_line()).then([this, &req] {
+        return req.write_request_headers(_write_buf).then([this, &req] {
+            return _write_buf.write("\r\n", 2);
+        });
+    });
+}
+
+future<reply> client::recv_reply() {
+    http_response_parser parser;
+    return do_with(std::move(parser), [this] (auto& parser) {
+        parser.init();
+        return _read_buf.consume(parser).then([this, &parser] {
+            if (parser.eof()) {
+                throw std::runtime_error("Invalid response");
+            }
+
+            auto resp = parser.get_parsed_response();
```

Comment:
I lost track - does this read the entire response (not a great idea) or does reply give you a way to read the response body incrementally?

Reply:
Parser stops after headers and leaves the whole body in the stream. Then the caller is supposed to get `client::in()` and read from the provided stream until its EOF